### ISSUE TITLE
Ncap2 and floating array with an unfinished number as a step

### ISSUE THREAD

**Issue by EnlNovius** [nco/nco#233]
Hello,
I encountered a strange behavior with the use of ncap2 for array generation. I don't know if this behavior is normal (caused by the rounding of unfinished binary values). If this behavior is "normal", I'm looking for the best solution to work around it.

The idea is to generate a new dimension "lat" with its associated variable "lat". My goal is to obtain the following variable :
lat(lat) : 7.95, 8.05, 8.15, 8.25, 8.35, 8.45, 8.55, 8.65, 8.75, 8.85, 8.95, 9.05, 9.15, 9.25, 9.35, 9.45, 9.55, 9.65, 9.75, 9.85, 9.95, 10.05, 10.15, 10.25, 10.35 ;

If I run : `ncap2 -s 'defdim("new_latitude", 25); new_latitude = array(7.95, 0.1, $new_latitude);' input.nc output.nc`

I get with `ncdump -v new_latitude output.nc` :

` >>> 7.95, 8.05, 8.15, 8.25, 8.35, 8.45, 8.55, 8.65, 8.75, 8.85, 
8.95, 9.05, 9.15, 9.25, 9.35, 9.45, 9.55, 9.65, 9.74999999999999, 
9.84999999999999, 9.94999999999999, 10.05, 10.15, 10.25, 10.35`
    
A functional solution is to first work with integers and then passing to float :
`ncap2 -s 'defdim("new_latitude", 25); new_latitude = array(795, 10, $new_latitude)/100.0f;' input.nc output.nc`

This code give the wanted result :
` >>> 7.95, 8.05, 8.15, 8.25, 8.35, 8.45, 8.55, 8.65, 8.75, 8.85, 
    8.95, 9.05, 9.15, 9.25, 9.35, 9.45, 9.55, 9.65, 9.75, 9.85, 9.95, 10.05, 
    10.15, 10.25, 10.35`
    
An interesting point is that by shifting the starting point of the array, the problem does not necessarily appear. For example executing : `ncap2 -s 'defdim("new_latitude", 25); new_latitude = array(8.95, 0.1, $new_latitude);' input.nc output.nc` gives the following result :
` >>> 8.95, 9.05, 9.15, 9.25, 9.35, 9.45, 9.55, 9.65, 9.75, 9.85, 
9.95, 10.05, 10.15, 10.25, 10.35, 10.45, 10.55, 10.65, 10.75, 10.85, 
10.95, 11.05, 11.15, 11.25, 11.35`

**Comment by czender** [comment]
Thank you for reporting this. It is "normal" in that you are using the `array` function as intended. Your workaround works well. 
@hmb1 this behavior makes me suspect that `array()` creates output values by adding the increment to the previous value. It should instead multiply the increment value by the index and add that to the initial value. Can you please verify whether my hypothesis is correct? If so, please fix it. If not, then I do not understand this behavior either.

**Comment by EnlNovius** [comment]
I wonder if the problem is not partly due to the display through ncdump -v.

**Comment by EnlNovius** [comment]
I tried to read the variable through python :

```
>>> import netCDF4 as nc
>>> import decimal
>>> file = nc.Dataset("output.nc")
>>> file['new_latitude'][:]
masked_array(data=[ 7.95,  8.05,  8.15,  8.25,  8.35,  8.45,  8.55,  8.65,
                    8.75,  8.85,  8.95,  9.05,  9.15,  9.25,  9.35,  9.45,
                    9.55,  9.65,  9.75,  9.85,  9.95, 10.05, 10.15, 10.25,
                   10.35],
             mask=False,
       fill_value=1e+20)
>>> import decimal
>>> [decimal.Decimal.from_float(v) for v in file['new_latitude'][:]]
[Decimal('7.95000000000000017763568394002504646778106689453125'), 
Decimal('8.050000000000000710542735760100185871124267578125'), 
Decimal('8.1500000000000003552713678800500929355621337890625'), 
Decimal('8.25'), 
Decimal('8.3499999999999996447286321199499070644378662109375'), 
Decimal('8.449999999999999289457264239899814128875732421875'), 
Decimal('8.5499999999999989341858963598497211933135986328125'), 
Decimal('8.64999999999999857891452847979962825775146484375'), 
Decimal('8.7499999999999982236431605997495353221893310546875'), 
Decimal('8.849999999999997868371792719699442386627197265625'), 
Decimal('8.9499999999999975131004248396493494510650634765625'), 
Decimal('9.0499999999999971578290569595992565155029296875'), 
Decimal('9.1499999999999968025576890795491635799407958984375'), 
Decimal('9.249999999999996447286321199499070644378662109375'), 
Decimal('9.3499999999999960920149533194489777088165283203125'), 
Decimal('9.44999999999999573674358543939888477325439453125'), 
Decimal('9.5499999999999953814722175593487918376922607421875'), 
Decimal('9.649999999999995026200849679298698902130126953125'), 
Decimal('9.7499999999999946709294817992486059665679931640625'), 
Decimal('9.849999999999994315658113919198513031005859375'), 
Decimal('9.9499999999999939603867460391484200954437255859375'), 
Decimal('10.049999999999993605115378159098327159881591796875'), 
Decimal('10.1499999999999932498440102790482342243194580078125'), 
Decimal('10.24999999999999289457264239899814128875732421875'), 
Decimal('10.3499999999999925393012745189480483531951904296875')]
```

**Comment by hmb1** [comment]
@czender
your diagnosis is correct.  
Have modified  the template code file.  
see new branch

**Comment by czender** [comment]
Thanks, @hmb1. I merged that. If you suspect that the replaced algorithm, which was suseptible to rounding error, is in use elsewhere in the code (like in other ncap2 functions), please replace it with the more accurate algorithm as well. @EnlNovius we think we have identified and fixed the problem in the latest snapshot. If you can, please build and test and let us know (by reopening this PR) if this does not solve the problem. Otherwise this will be released in NCO 4.9.8 within a week or two.
